cmd/query/main.go

```diff
@@ -89,9 +103,27 @@ func main() {
 			sHandler.RegisterRoutes(r)
 			portStr := ":" + strconv.Itoa(queryOpts.QueryPort)
 			recoveryHandler := recoveryhandler.NewRecoveryHandler(logger, true)
-			logger.Info("Starting jaeger-query HTTP server", zap.Int("port", queryOpts.QueryPort))
-			if err := http.ListenAndServe(portStr, recoveryHandler(r)); err != nil {
-				logger.Fatal("Could not launch service", zap.Error(err))
+
+			go func() {
+				// there's still a chance that a request comes in between the `ready` flag and the server's
+				// actual startup, but this would be a rare situation, as we assume the `ListenAndServe` is fast enough
+				// between allocating a port and serving a request.
+				hcChannel <- http.StatusNoContent
+
+				logger.Info("Starting jaeger-query HTTP server", zap.Int("port", queryOpts.QueryPort))
+				if err := http.ListenAndServe(portStr, recoveryHandler(r)); err != nil {
```

Comment:
there is no bullet-proof way to avoid that aside from having the healthcheck to actually ping the other ports opened by the service. Which could be a future enhancement of it, e.g. by passing "ping actors" to the Ready() functions, e.g.

```go
hc.Ready(
    healthcheck.HTTPChecker(apiPort),
    healthcheck.HTTPChecker(zipkinPort),
    healthcheck.TChannelChecker(collectorPort),
)
```